docs/JAX_machine_translation.md

```diff
@@ -0,0 +1,502 @@
+                                  in_features=embed_dim,
+                                  decode=False,
+                                  rngs=rngs)
+
+        self.dense_proj = nnx.Sequential(
+            *[
+                nnx.Linear(embed_dim, latent_dim, rngs=rngs),
+                nnx.relu,
+                nnx.Linear(latent_dim, embed_dim, rngs=rngs),
+            ]
+        )
+        self.layernorm_1 = nnx.LayerNorm(embed_dim, rngs=rngs) ## check inputs
+        self.layernorm_2 = nnx.LayerNorm(embed_dim, rngs=rngs)
+        self.layernorm_3 = nnx.LayerNorm(embed_dim, rngs=rngs)
+
+    def __call__(self, inputs, encoder_outputs, mask=None):
+        causal_mask = self.get_causal_attention_mask(inputs.shape[1])
+        if mask is not None:
+            padding_mask = jnp.expand_dims(mask, axis=1).astype(jnp.int32)
+            padding_mask = jnp.minimum(padding_mask, causal_mask)
+        else:
+            padding_mask = None
+        attention_output_1 = self.attention_1(
+            inputs_q=inputs, inputs_v=inputs, inputs_k=inputs,  mask=causal_mask  ## check kwargs, specifically qkv
+        )
+        out_1 = self.layernorm_1(inputs + attention_output_1)
+
+        attention_output_2 = self.attention_2( ## https://github.com/google/flax/blob/main/flax/nnx/nn/attention.py#L403-L405
+            inputs_q=out_1,
+            inputs_v=encoder_outputs,
+            inputs_k=encoder_outputs,
+            mask=padding_mask,
+        )
+        out_2 = self.layernorm_2(out_1 + attention_output_2)
+
+        proj_output = self.dense_proj(out_2)
+        return self.layernorm_3(out_2 + proj_output)
+
+    def get_causal_attention_mask(self, sequence_length):
+        i = jnp.arange(sequence_length)[:, None]
+        j = jnp.arange(sequence_length)
+        mask = (i >= j).astype(jnp.int32)
+        mask = jnp.reshape(mask, (1, 1, sequence_length, sequence_length))
+        return mask
+```
+
+## Build out Data Loader, Model and Training Definitions
+It can be more computationally efficient to use pygrain for the data load stage, but this way it's abundandtly clear what's happening: data pairs go in and sets of jnp arrays come out, in step with our original dictionaries. 'Encoder_inputs', 'decoder_inputs' and 'target_output'.
+
+```{code-cell} ipython3
+class DataLoader:
+    def __init__(self, data, batch_size):
+        self.data = data
+        self.batch_size = batch_size
+        self.num_batches = len(data) // batch_size
+
+    def __iter__(self):
+        for i in range(self.num_batches):
+            batch = self.data[i * self.batch_size: (i + 1) * self.batch_size]
+            encoder_inputs = [item["encoder_inputs"] for item in batch]
+            decoder_inputs = [item["decoder_inputs"] for item in batch]
+            target_output = [item["target_output"] for item in batch]
+            yield (jnp.array(encoder_inputs), jnp.array(decoder_inputs), jnp.array(target_output))
+```
```

Comment:
Let's use grain for the data loading instead of self crafted class as IMO we have to show best practices such that reader can take this code and reuse it for their own use-cases with larger datasets etc.